### PR TITLE
MAT-1301 Use Bor root-hash API

### PR DESCRIPTION
**Pull request by ptsayli**

### CODE REVIEW

File: checkpoint/querier.go [hunk 1]
```diff
@@ -141,7 +141,7 @@ func handleQueryNextCheckpoint(ctx sdk.Context, req abci.RequestQuery, keeper Ke
 	params := keeper.GetParams(ctx)
 	end := start + params.AvgCheckpointLength
 
-	rootHash, err := types.GetHeaders(start, end, params.MaxCheckpointLength)
+	rootHash, err := contractCaller.GetRootHash(start, end, params.MaxCheckpointLength)
 	if err != nil {
 		return nil, sdk.ErrInternal(sdk.AppendMsgToErr(fmt.Sprintf("could not fetch headers for start:%v end:%v error:%v", start, end, err), err.Error()))
```
Comment (mankenavenkatesh) [inline checkpoint/querier.go:146]: Please change the error message.

File: go.mod [hunk 2]
```diff
@@ -54,13 +46,10 @@ require (
 	github.com/tendermint/tendermint v0.32.7
 	github.com/tendermint/tm-db v0.2.0
 	github.com/tyler-smith/go-bip39 v1.0.2 // indirect
-	github.com/wsddn/go-ecdh v0.0.0-20161211032359-48726bab9208 // indirect
-	github.com/xsleonard/go-merkle v0.0.0-20190513010722-059a904a0479
+	github.com/xsleonard/go-merkle v1.1.0
+	golang.org/x/crypto v0.0.0-20200429183012-4b2356b1ed79 // indirect
```
Comment (mankenavenkatesh) [inline go.mod:50]: why package changes?

File: helper/call.go [hunk 3]
```diff
@@ -257,6 +261,30 @@ func (c *ContractCaller) GetHeaderInfo(headerID uint64, rootChainInstance *rootc
...
+func (c *ContractCaller) GetRootHash(start uint64, end uint64, checkpointLength uint64) ([]byte, error) {
+	var rootHash []byte
+
+	noOfBlock := end - start + 1
+
+	if noOfBlock > checkpointLength {
+		return nil, errors.New("number of headers requested exceeds")
+	}
+
+	if start > end {
```
Comment (mankenavenkatesh) [inline helper/call.go:274]: pls move this condition to line num 267

File: checkpoint/client/rest/query.go [hunk 4]
```diff
@@ -208,9 +208,10 @@ func checkpointHandlerFn(cliCtx context.CLIContext) http.HandlerFunc {
 		}
 		var params types.Params
 		json.Unmarshal(res, &params)
+		contractCallerObj, err := helper.NewContractCaller()
 
 		// get headers
-		roothash, err := types.GetHeaders(uint64(start), uint64(end), params.MaxCheckpointLength)
+		roothash, err := contractCallerObj.GetRootHash(uint64(start), uint64(end), params.MaxCheckpointLength)
 		if err != nil {
 			RestLogger.Error("Unable to get header", "Start", start, "End", end, "Error", err)
```
Comment (mankenavenkatesh) [inline checkpoint/client/rest/query.go:216]: Pls change the log

**Comment by jdkanani** [comment]
@ptsayli Needs to use eth client. Changing status.

File: helper/call.go [hunk 5]
```diff
@@ -257,6 +261,30 @@ func (c *ContractCaller) GetHeaderInfo(headerID uint64, rootChainInstance *rootc
...
+		return nil, errors.New("start is greater than end")
+	}
+
+	startBlockNumber := "0x" + strconv.FormatUint(start, 16)
+	endBlockNumer := "0x" + strconv.FormatUint(end, 16)
+
+	if err := c.MaticRPCClient.CallContext(context.Background(), &rootHash, "eth_getRootHash", startBlockNumber, endBlockNumer); err != nil {
+		return nil, err
```
Comment (mankenavenkatesh) [inline helper/call.go:282]: Pls add Error log here. Helps in knowing if error is with MaticClient or not.

File: go.mod (hunk 2 above)
Comment (ptsayli) [reply go.mod:50]: Actually that are auto updated with new bor version, also giving warnings on build
I had talked to @atvanguard regarding warnings

File: helper/call.go [hunk 6]
```diff
@@ -76,6 +77,7 @@ type ContractCaller struct {
 	MainChainClient  *ethclient.Client
 	MainChainRPC     *rpc.Client
 	MaticChainClient *ethclient.Client
+	MaticRPCClient   *rpc.Client
```
Comment (jdkanani) [inline helper/call.go:80]: Remove unwanted client object?

**Pull request merged**